httpcore/_backends/curio.py

```diff
@@ -0,0 +1,214 @@
+import select
+import socket
+from ssl import SSLContext, SSLSocket
+from typing import Dict, Optional, Type, Union
+
+import curio
+import curio.io
+
+from .._exceptions import (
+    ConnectError,
+    ConnectTimeout,
+    ReadError,
+    ReadTimeout,
+    TimeoutException,
+    WriteError,
+    WriteTimeout,
+    map_exceptions,
+)
+from .._types import TimeoutDict
+from .._utils import get_logger
+from .base import AsyncBackend, AsyncLock, AsyncSemaphore, AsyncSocketStream
+
+logger = get_logger("curio_backend")
+
+one_day_in_seconds = 60 * 60 * 24
+
+
+async def wrap_ssl_client(
+    sock: curio.io.Socket,
+    ssl_context: SSLContext,
+    server_hostname: bytes,
+) -> curio.io.Socket:
+    kwargs = {
+        "server_hostname": server_hostname,
+        "do_handshake_on_connect": sock._socket.gettimeout() != 0.0,
+    }
+
+    socket = curio.io.Socket(ssl_context.wrap_socket(sock._socket, **kwargs))
+    await socket.do_handshake()
+
+    return socket
+
+
+def convert_timeout(value: Optional[float]) -> int:
+    return int(value) if value is not None else one_day_in_seconds
```

Comment:
(Nit)

Plus, I suppose we're using "a very long timeout" because Curio doesn't support passing `float("inf")` — or does it?

```suggestion
    return int(value) if value is not None else ONE_DAY_IN_SECONDS
```

Reply:
Once I tried to do it like in `trio` (treat `None` as `float("inf")`, I stuck with the following problem:
<details>
<summary>Error log for float("inf")</summary>
    ```
self = <selectors.KqueueSelector object at 0x10f2ef310>, timeout = inf

    def select(self, timeout=None):
        timeout = None if timeout is None else max(timeout, 0)
        max_ev = len(self._fd_to_key)
        ready = []
        try:
>           kev_list = self._selector.control(None, max_ev, timeout)
E           TypeError: timeout argument must be a number or None, got float
    ```
</details>

Once I tried to keep `None` as `None`, I got a `curio` error:
<details>
<summary>Error log for None</summary>
    ```
self = <curio.time._TimeoutAfter object at 0x104031520>, ty = None, val = None
tb = None

    async def __aexit__(self, ty, val, tb):
        current_clock = await _unset_timeout(self._prior)
    
        # Discussion.  If a timeout has occurred, it will either
        # present itself here as a TaskTimeout or TimeoutCancellationError
        # exception.  The value of this exception is set to the current
        # kernel clock which can be compared against our own deadline.
        # What happens next is driven by these rules:
        #
        # 1.  If we are the outer-most context where the timeout
        #     period has expired, then a TaskTimeout is raised.
        #
        # 2.  If the deadline has expired for at least one outer
        #     context, (but not us), a TimeoutCancellationError is
        #     raised.  This means that time has expired elsewhere.
        #     We're being cancelled because of that, but the reason
        #     for the cancellation wasn't due to a timeout on our
        #     part.
        #
        # 3.  If the timeout period has not expired on ANY remaining
        #     timeout context, it means that a timeout has escaped
        #     some inner timeout context where it should have been
        #     caught. This is an operational error.  We raise
        #     UncaughtTimeoutError.
    
        try:
            if ty in (TaskTimeout, TimeoutCancellationError):
                timeout_clock = val.args[0]
                # Find the outer most deadline that has expired
                for n, deadline in enumerate(self._deadlines):
                    if deadline <= timeout_clock:
                        break
                else:
                    # No remaining context has expired. An operational error
                    raise UncaughtTimeoutError('Uncaught timeout received')
    
                if n < len(self._deadlines) - 1:
                    if ty is TaskTimeout:
                        raise TimeoutCancellationError(val.args[0]).with_traceback(tb) from None
                    else:
                        return False
                else:
                    # The timeout is us.  Make sure it's a TaskTimeout (unless ignored)
                    self.result = self._timeout_result
                    self.expired = True
                    if self._ignore:
                        return True
                    else:
                        if ty is TimeoutCancellationError:
                            raise TaskTimeout(val.args[0]).with_traceback(tb) from None
                        else:
                            return False
            elif ty is None:
>               if current_clock > self._deadlines[-1]:
E               TypeError: '>' not supported between instances of 'float' and 'NoneType'
    ```
</details>

Moreover, I tried to find a maximum value of timeout, which do not cause an error. The value I found is `2**26-1` == `67108863`. I thought about that and decided to choose any significant time interval less than `2**26-1`, for example one day

Reply:
Yes, I know eg trio doesn't allow `None` and requires users to pass `float("inf")` to mean "don't ever timeout", so I was curious if this was supported by `curio`, so that we'd do:

```python
    return float("inf") if value is None else int(value)

```

Also, wait… Should this be `float(value)`? 🤔🤔🤔